test/tree.spec.js

```diff
@@ -523,4 +523,39 @@ describe('Tree Functions', () => {
     await walkAsyncTest
     expect(tree.a.b.c.length).to.equal(4)
   })
+  it('mapTreeLeaves', () => {
+    let tree = { a: { b: { c: [1, 2, 3] } } }
+    let double = x => x * 2
+    let result = F.mapTreeLeaves()(double, tree)
+    expect(tree).to.deep.equal({ a: { b: { c: [1, 2, 3] } } })
+    expect(result).to.deep.equal({ a: { b: { c: [2, 4, 6] } } })
+  })
+  it('mapTreeLeaves contexture tree', () => {
+    let tree = {
+      key: 'root',
+      children: [
+        { key: 'criteria', children: [{ key: 'filter' }, { key: 'f2' }] },
+        { key: 'analysis', children: [{ key: 'results' }] },
+      ],
+    }
+    let getChildren = x => x.children
+    let writeChild = (node, index, [parent]) => {
+      parent.children[index] = node
+    }
+    let mapLeaves = F.mapTreeLeaves(getChildren, writeChild)
```

Comment:
Could we make the API more symmetrical by having `getChildren` and `setChildren` where the latter would take `(parent, children)` and return `parent` with the new `children` set?

Reply:
The challenge is that you don’t want to override all of the children, only the ones that are necessary. Also, you can assume children are an array - some tree structures have object children (for example, JSON schemas). If you take all the children and return them, the writing method also needs to understand how to traverse the current children

Reply:
However this is a good point, we should have tests for those use cases

Reply:
> The challenge is that you don’t want to override all of the children, only the ones that are necessary

Makes sense

> If you take all the children and return them, the writing method also needs to understand how to traverse the current children

Right. This is why whenever I think of mapping over a tree, I think of a post-order walk, not a pre-order one.